apps/services/user-notification/src/app/modules/notifications/notifications.service.ts

```diff
@@ -151,4 +163,135 @@ export class NotificationsService {
     }
     return template
   }
+
+  async findOne(
+    user: User,
+    id: number,
+    locale: string,
+  ): Promise<RenderedNotificationDto> {
+    // Retrieve the notification
+    const notification = await this.notificationModel.findOne({
+      where: {
+        id: id,
+        recipient: user.nationalId,
+      },
+    })
+
+    if (!notification) {
+      throw new NotFoundException(
+        'Notification not found or does not belong to the user',
+      )
+    }
+
+    // Process and format the notification
+    try {
+      const template = await this.getTemplate(notification.templateId, locale)
+      const formattedTemplate = this.formatArguments(notification, template)
+      return {
+        id: notification.id,
+        messageId: notification.messageId,
+        title: formattedTemplate.notificationTitle,
+        body: formattedTemplate.notificationBody,
+        dataCopy: formattedTemplate.notificationDataCopy,
+        clickAction: formattedTemplate.clickAction,
+        created: notification.created,
+        updated: notification.updated,
+        status: notification.status,
+      }
+    } catch (error) {
+      this.logger.error('Error formatting notification:', error)
+      throw new InternalServerErrorException('Error processing notification')
+    }
+  }
+
+  async findMany(user: User, query: any): Promise<PaginatedNotificationDto> {
+    const templates = await this.getTemplates(query.locale)
+    const paginatedListResponse = await paginate({
+      Model: this.notificationModel,
+      limit: query.limit || 10,
+      after: query.after,
+      before: query.before,
+      primaryKeyField: 'id',
+      orderOption: [['id', 'DESC']],
+      where: { recipient: user.nationalId },
+    })
+    const formattedNotifications = paginatedListResponse.data.map(
+      (notification) => {
+        const template = templates.find(
+          (template) => template.templateId === notification.templateId,
+        )
+        if (template) {
+          try {
+            const formattedTemplate = this.formatArguments(
+              notification,
+              template,
+            )
+            return {
+              id: notification.id,
+              messageId: notification.messageId,
+              title: formattedTemplate.notificationTitle,
+              body: formattedTemplate.notificationBody,
+              dataCopy: formattedTemplate.notificationDataCopy,
+              clickAction: formattedTemplate.clickAction,
+              created: notification.created,
+              updated: notification.updated,
+              status: notification.status,
+            }
```

Comment:
This pattern is repeated 3x in this service and two of those also repeat `getTemplate`. We should create a private helper function for this. It can accept templateId or template, if the template is provided it skips getting the template, then formats and maps.